presto-spi/src/main/java/com/facebook/presto/spi/block/StructBuilder.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.Slices;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+
+public class StructBuilder
+{
+    private final BlockBuilderStatus builderStatus;
+    private final BlockBuilder builder;
+    private final List<Type> parameterTypes;
+    private final Iterator<Type> typeIterator;
+
+    public static StructBuilder arrayBuilder(Type elementType)
+    {
+        return new StructBuilder(asList(elementType));
+    }
+
+    public static StructBuilder mapyBuilder(Type keyType, Type valueType)
+    {
+        return new StructBuilder(asList(keyType, valueType));
+    }
+
+    public static StructBuilder rowBuilder(List<Type> parameterTypes)
+    {
+        return new StructBuilder(parameterTypes);
+    }
+
+    public static StructBuilder builder(int expectedSize, List<Type> parameterTypes)
+    {
+        return new StructBuilder(expectedSize, parameterTypes);
+    }
+
+    private StructBuilder(List<Type> parameterTypes)
+    {
+        this(1024, parameterTypes);
+    }
+
+    StructBuilder(int expectedSize, List<Type> parameterTypes)
+    {
+        this(new BlockBuilderStatus(expectedSize, expectedSize), parameterTypes);
+    }
+
+    private StructBuilder(BlockBuilderStatus builderStatus, List<Type> parameterTypes)
+    {
+        if (parameterTypes.isEmpty()) {
+            throw new IllegalArgumentException("parameterTypes are empty");
+        }
+
+        this.builderStatus = builderStatus;
+        this.builder = new VariableWidthBlockBuilder(builderStatus);
+        this.parameterTypes = unmodifiableList(parameterTypes);
+        this.typeIterator = new CycleIterator();
+    }
+
+    public static Block readBlock(SliceInput input)
+    {
+        return new VariableWidthBlockEncoding().readBlock(input);
+    }
+
+    public Slice build()
+    {
+        int size = builderStatus.getSizeInBytes() + 5 * (builder.getPositionCount() + 1);
```

Comment:
I've added a short description of how `VariableWidthEncoding` serialize into bytes.